File: kernel/logging/records/queue.go
package records

import (
	"container/list"
	"sync"
)

// Frame is already bounded and encoded for its next hop. The queue takes
// ownership of Payload; producers must not mutate it after admission.
type Frame struct {
	Payload       []byte
	Source, Level string
}
type queuedFrame struct {
	Frame
	order uint64
}

type QueueStats struct {
	PendingBytes   int       `json:"pending_bytes"`
	PendingRecords int       `json:"pending_records"`
	Dropped        [4]uint64 `json:"dropped"`
}

// Queue accounts for both waiting and currently consumed batches. A stalled
// socket/file write cannot release credit and hide another unbounded buffer.
// It reserves bytes and record slots for warnings/errors and evicts waiting
// low-severity records before dropping a newly arriving warning/error.
type Queue struct {
	mu                      sync.Mutex
	low, high               list.List
	maximum, reserve, slots int
	stats                   QueueStats
	order                   uint64
	notify                  chan struct{}
}

func NewQueue(maximum, reserve, slots int) *Queue {
	if maximum < MaxRecord || reserve < 0 || reserve >= maximum || slots < 4 {
		panic("invalid logging queue bounds")
	}
	return &Queue{maximum: maximum, reserve: reserve, slots: slots, notify: make(chan struct{}, 1)}
}

func (q *Queue) Notify() <-chan struct{} { return q.notify }

func (q *Queue) Add(frame Frame) bool {
	q.mu.Lock()
	defer q.mu.Unlock()
	level := Severity(frame.Level)
	if level < 0 || len(frame.Payload) == 0 || len(frame.Payload) > MaxFrame {
		return false
	}
	weight := len(frame.Payload) + 4
	limit, slots := q.maximum, q.slots
	if level < 2 {
		limit -= q.reserve
		slots -= max(1, q.slots/4)
	} else {
		for (q.stats.PendingBytes+weight > limit || q.stats.PendingRecords >= slots) && q.low.Len() > 0 {
			e := q.low.Front()
			old := e.Value.(queuedFrame)
			q.low.Remove(e)
			q.stats.PendingBytes -= len(old.Payload) + 4
			q.stats.PendingRecords--
			q.stats.Dropped[Severity(old.Level)]++
		}
	}
	if q.stats.PendingBytes+weight > limit || q.stats.PendingRecords >= slots {
		q.stats.Dropped[level]++
		return false
	}
	q.order++
	item := queuedFrame{Frame: frame, order: q.order}
	if level < 2 {
		q.low.PushBack(item)
	} else {
		q.high.PushBack(item)
	}
	q.stats.PendingBytes += weight
	q.stats.PendingRecords++
	select {
	case q.notify <- struct{}{}:
	default:
	}
	return true
}

// Batch has one consumer; Release must be called after the downstream operation
// finishes, including failures. Internal ordering is not persisted in records.
type Batch struct {
	Frames   []Frame
	bytes    int
	queue    *Queue
	released bool
}

func (q *Queue) Take(maximum int) *Batch {
	q.mu.Lock()
	defer q.mu.Unlock()
	b := &Batch{queue: q}
	for {
		low, high := q.low.Front(), q.high.Front()
		var selected *list.Element
		owner := &q.low
		if low == nil {
			selected = high
			owner = &q.high
		} else if high == nil || low.Value.(queuedFrame).order < high.Value.(queuedFrame).order {
			selected = low
		} else {
			selected = high
			owner = &q.high
		}
		if selected == nil {
			break
		}
		item := selected.Value.(queuedFrame)
		weight := len(item.Payload) + 4
		if b.bytes+weight > maximum {
			break
		}
		owner.Remove(selected)
		b.Frames = append(b.Frames, item.Frame)
		b.bytes += weight
	}
	return b
}

func (b *Batch) Release() {
	q := b.queue
	q.mu.Lock()
	defer q.mu.Unlock()
	if b.released {
		return
	}
	b.released = true
	q.stats.PendingBytes -= b.bytes
	q.stats.PendingRecords -= len(b.Frames)
	clear(b.Frames)
	b.Frames = nil
	if q.low.Len()+q.high.Len() > 0 {
		select {
		case q.notify <- struct{}{}:
		default:
		}
	}
}

func (q *Queue) Stats() QueueStats { q.mu.Lock(); defer q.mu.Unlock(); return q.stats }

// Drop accounts for losses after admission (for example a failed socket write).
func (q *Queue) Drop(level string, count uint64) {
	q.mu.Lock()
	defer q.mu.Unlock()
	if n := Severity(level); n >= 0 {
		q.stats.Dropped[n] += count
	}
}

File: defaults/config/runtime/deno/logging/queue.ts
import {
  addDrops,
  type DropCounts,
  type Level,
  LEVELS,
  MAX_FRAME,
} from "./protocol.ts";

interface Entry {
  frame: Uint8Array;
  level: Level;
  order: number;
}

class Ring {
  #items: Array<Entry | undefined>;
  #head = 0;
  length = 0;
  constructor(capacity: number) {
    this.#items = new Array(capacity);
  }
  peek(): Entry | undefined {
    return this.#items[this.#head];
  }
  push(entry: Entry): void {
    this.#items[(this.#head + this.length++) % this.#items.length] = entry;
  }
  shift(): Entry | undefined {
    if (this.length === 0) return undefined;
    const entry = this.#items[this.#head];
    this.#items[this.#head] = undefined;
    this.#head = (this.#head + 1) % this.#items.length;
    this.length--;
    return entry;
  }
}

// Credit includes frames held by the one socket write in flight. Reserved high
// priority space cannot be consumed by low severity traffic.
export class LogQueue {
  readonly maximum: number;
  readonly reserve: number;
  readonly slots: number;
  readonly dropped: DropCounts = [0, 0, 0, 0];
  #low: Ring;
  #high: Ring;
  #order = 0;
  bytes = 0;
  records = 0;

  constructor(maximum = 128 * 1024, reserve = 32 * 1024, slots = 512) {
    if (
      maximum < MAX_FRAME + 132 || reserve < 0 || reserve >= maximum ||
      slots < 4
    ) {
      throw new Error("invalid log queue bounds");
    }
    this.maximum = maximum;
    this.reserve = reserve;
    this.slots = slots;
    this.#low = new Ring(slots);
    this.#high = new Ring(slots);
  }

  get waiting(): number {
    return this.#low.length + this.#high.length;
  }

  drop(level: Level, count = 1): void {
    const counts: DropCounts = [0, 0, 0, 0];
    counts[LEVELS.indexOf(level)] = count;
    addDrops(this.dropped, counts);
  }

  add(frame: Uint8Array, level: Level): boolean {
    if (
      frame.length < 5 || frame.length > MAX_FRAME + 4 ||
      !LEVELS.includes(level)
    ) return false;
    const high = LEVELS.indexOf(level) >= 2;
    const limit = this.maximum - (high ? 0 : this.reserve);
    const slots = this.slots -
      (high ? 0 : Math.max(1, Math.floor(this.slots / 4)));
    const weight = frame.length + 128;
    if (high) {
      while (
        (this.bytes + weight > limit || this.records >= slots) &&
        this.#low.length > 0
      ) {
        const old = this.#low.shift()!;
        this.bytes -= old.frame.length + 128;
        this.records--;
        this.drop(old.level);
      }
    }
    if (this.bytes + weight > limit || this.records >= slots) {
      this.drop(level);
      return false;
    }
    (high ? this.#high : this.#low).push({
      frame,
      level,
      order: ++this.#order,
    });
    this.bytes += weight;
    this.records++;
    return true;
  }

  take(maximum: number): { entries: Entry[]; release(): void } {
    const entries: Entry[] = [];
    let length = 0;
    let weight = 0;
    while (this.waiting > 0) {
      const low = this.#low.peek(), high = this.#high.peek();
      const ring =
        low === undefined || (high !== undefined && high.order < low.order)
          ? this.#high
          : this.#low;
      const entry = ring.peek()!;
      if (length + entry.frame.length > maximum) break;
      ring.shift();
      entries.push(entry);
      length += entry.frame.length;
      weight += entry.frame.length + 128;
    }
    let released = false;
    return {
      entries,
      release: () => {
        if (released) return;
        released = true;
        this.bytes -= weight;
        this.records -= entries.length;
        entries.length = 0;
      },
    };
  }
}
